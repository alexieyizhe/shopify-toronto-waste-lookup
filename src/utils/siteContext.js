import React from 'react';

export const wasteItems = [
  {
    title: 'Garbage (wrapping and tying)',
    body: [
      'Place item in the <strong>Garbage Bin.</strong>'
    ]
  },
  {
    title: 'Garbage (tying)',
    body: [
      'Place item in the <strong>Garbage (tying) Bin.</strong>'
    ]
  },
  {
    title: 'Garbage (wrapping) ',
    body: [
      'Place item in the <strong>Garbage (wrapping) Bin.</strong>'
    ]
  }
];

export const ItemsContext = React.createContext({
  items: wasteItems,
  favs: new Set(), // TODO: make this into a set
  updateFavs: () => {}
});
