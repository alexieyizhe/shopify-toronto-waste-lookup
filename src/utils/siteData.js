// should refactor into json and move into data folder later

export const siteTheme = {
  colors: {
    primary: '#1d5a94',
    secondary: '#23975e',
    secondaryBg: '#b5d2c4',

    warning: '#F4A867',
    error: '#F07285',

    offBlack: '#333'
  },
  styling: {
    bodySpacing: '1vw',
  }
};

export const siteTitle = 'Toronto Waste Lookup';

export const searchBarEmptyPlaceholder = 'Enter a search';

export const searchResultsPlaceholders = {
  empty: {text: 'No results found. Try changing your search?', color: siteTheme.colors.warning},
  ready: {text: 'Search for a waste item above!', color: siteTheme.colors.secondary},
  waiting: {text: 'Fetching latest data from the Toronto Waste Wizard...', color: 'grey'},
  error: {text: 'An error occurred while fetching data. Please try refreshing the page.', color: siteTheme.colors.error},
  searching: {text: 'Searching for matching items...', color: siteTheme.colors.secondary},
}

export const favsEmptyPlaceholder = 'You have no favourites. Try adding some with the star icon next to a waste item\'s name!';

export const siteSubtitle = 'Favourites';

export const wasteDataAPIEndPoint = 'https://secure.toronto.ca/cc_sr_v1/data/swm_waste_wizard_APR?limit=1000';

export const FetchStateEnum = { ERROR: 0, WAITING: 1, READY: 2, SEARCHING: 3 };

export const LocalStorageKeysEnum = {
  lastAPIResults: 'wasteItems',
  lastSearchQuery: 'searchQuery',
  lastSearchResults: 'searchResults',
  lastFavourites: 'currentFavs',
  lastStatus: 'appStatus',
};
