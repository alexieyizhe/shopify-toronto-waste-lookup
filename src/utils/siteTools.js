import { css } from 'styled-components';

const displaySizes = {
  large: 1024,
  mid: 832,
  small: 640
};

// Create media template for responsive designs
export const mediaSize = Object.keys(displaySizes).reduce((acc, label) => {
  acc[label] = (...args) => css`
    @media (min-width: ${displaySizes[label] / 16}em) {
      ${css(...args)};
    }
  `;

  return acc;
}, {});
