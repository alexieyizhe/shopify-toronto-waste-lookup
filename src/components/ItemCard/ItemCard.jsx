import React, { forwardRef } from 'react';
import styled, { withTheme } from 'styled-components';
import posed from 'react-pose';

import { ItemsContext } from '../../utils/siteContext';
import { mediaSize } from '../../utils/siteTools';

const CardAnim = {
  enter: {
    x: 0,
    opacity: 1,
    delay: ({ delayIndex }) => delayIndex * 25
  },
  exit: { x: 100, opacity: 0 }
};

const ComponentContainer = styled.div`
  width: 100%;
  height: auto;
  margin-bottom: 2vw;

  display: flex;
  flex-direction: row;
  justify-content: space-evenly;

  ${mediaSize.tablet`
    width: 90%;
    padding: 5vw 2vw;
    box-shadow: rgba(0, 0, 0, 0.1) 0px 7px 40px;
    margin: 0 auto 5vw auto;

    display: grid;
    grid-template-columns: 1fr 9fr;
    grid-template-rows: auto auto;
    grid-template-areas:
      'fav title'
      'desc desc';
    justify-items: center;
    align-items: center;
  `}
`;

const FavouriteButton = styled.div`
  width: 3%;

  ${mediaSize.tablet`
    grid-area: fav;
    width: auto;
  `}

  ${mediaSize.mobile`
    align-self: start;
    padding-top: 0;
  `}


  & svg {
    cursor: pointer;

    transition: fill 200ms ease-in-out;
    fill: ${props => props.theme.colors.offBlack};

    &:hover, &.favourited {
      fill: ${props => props.theme.colors.secondary};
    }

    width: 1.5vw;
    height: 1.5vw;

    ${mediaSize.tablet`
      width: 3vw;
      height: 3vw;
    `}

    ${mediaSize.mobile`
      width: 4vw;
      height: 4vw;
    `}
  }
`;

const TitleContainer = styled.div`
  color: ${props => props.theme.colors.offBlack};
  width: 45%;

  ${mediaSize.tablet`
    grid-area: title;
    width: 100%;

    font-weight: 600; 
  `}
`;

const DescContainer = styled.div`
  color: ${props => props.theme.colors.offBlack};
  width: 50%;

  & > ul {
    margin: 0;
    padding: 0;
  }

  ${mediaSize.tablet`
    grid-area: desc;
    justify-self: center;
    width: 95%;

    & > ul > li {
      list-style-type: none;
      margin-top: 2vw;
    }
  `}

  ${mediaSize.mobile`

  `}
`;

// HTML needs to be unescaped, otherwise React treats it as regular ol' text
// Code source: https://stackoverflow.com/questions/22279231/using-js-jquery-how-can-i-unescape-html-and-put-quotes-back-in-the-str
const unescapeHtml = (safe) => (
  safe.replace(/&amp;/g, '&')
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>')
      .replace(/&quot;/g, '"')
      .replace(/&#039;/g, "'")
);

// Component
export const NoAnimItemCard = forwardRef(({title, body, ith, isFavourite, theme}, innerRef) => (
  <ItemsContext.Consumer>
    {({items, favs, updateFavs}) => (
      <ComponentContainer ref={innerRef}>
        <FavouriteButton onClick={() => updateFavs(ith, isFavourite)}>
          {/* Icon source: https://www.s-ings.com/typicons/ */}
          <svg width="1.5vw" height="1.5vw" baseProfile="tiny" version="1.2" viewBox="0 0 24 24" className={isFavourite ? 'favourited' : ''}>
            <path d="M3.1 11.3l3.6 3.3-1 4.6c-.1.6.1 1.2.6 1.5.2.2.5.3.8.3.2 0 .4 0 .6-.1 0 0 .1 0 .1-.1l4.1-2.3 4.1 2.3s.1 0 .1.1c.5.2 1.1.2 1.5-.1.5-.3.7-.9.6-1.5l-1-4.6c.4-.3 1-.9 1.6-1.5l1.9-1.7.1-.1c.4-.4.5-1 .3-1.5s-.6-.9-1.2-1h-.1l-4.7-.5-1.9-4.3s0-.1-.1-.1c-.1-.7-.6-1-1.1-1-.5 0-1 .3-1.3.8 0 0 0 .1-.1.1L8.7 8.2 4 8.7h-.1c-.5.1-1 .5-1.2 1-.1.6 0 1.2.4 1.6z"/>
          </svg>
        </FavouriteButton>
        <TitleContainer>{title}</TitleContainer>
        <DescContainer dangerouslySetInnerHTML={{__html: unescapeHtml(body)}}></DescContainer> { /*  TODO: possibly implement sanitizing script that makes sure the only html elements are <ul>, <li>, and <strong> */ }
        { /*  ^^ Not really sure what the best way to sanitize this and make it safe without expending lots of effort with parsing the HTML body it returns, but the API is trusted so this should be safe */ }
      </ComponentContainer>
    )}
  </ItemsContext.Consumer>
));

export default posed(withTheme(NoAnimItemCard))(CardAnim);
