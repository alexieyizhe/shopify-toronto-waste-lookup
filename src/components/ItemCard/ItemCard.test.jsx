import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import { siteTheme } from '../../utils/siteData';

import ItemCard from './ItemCard';

/* eslint-disable no-undef */
it('renders correctly', () => {
  const tree = renderer.create(
    <ThemeProvider theme={siteTheme}>
      <ItemCard title="This is a test title" body={["This is a test body"]} />
    </ThemeProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
