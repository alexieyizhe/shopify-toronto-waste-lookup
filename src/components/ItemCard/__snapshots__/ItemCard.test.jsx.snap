// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c0 {
  width: 100%;
  height: auto;
  padding-bottom: 2vw;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  font-size: 1.5vw;
}

.c1 {
  grid-area: fav;
  width: 3%;
}

.c2 {
  grid-area: title;
  color: #333;
  width: 45%;
}

.c3 {
  grid-area: desc;
  color: #333;
  width: 50%;
}

.c3 > ul {
  margin: 0;
  padding: 0;
}

<div
  className="c0"
>
  <div
    className="c1"
    onClick={[Function]}
  >
    <svg
      fill="none"
      height="1vw"
      stroke="#333"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width="1vw"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polygon
        points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
      />
    </svg>
  </div>
  <div
    className="c2"
  >
    This is a test title
  </div>
  <div
    className="c3"
  >
    <ul>
      <li
        dangerouslySetInnerHTML={
          Object {
            "__html": "This is a test body",
          }
        }
      />
       
      <li
        dangerouslySetInnerHTML={
          Object {
            "__html": "This is a test body",
          }
        }
      />
       
      <li
        dangerouslySetInnerHTML={
          Object {
            "__html": "This is a test body",
          }
        }
      />
       
    </ul>
  </div>
</div>
`;
